### PR TITLE
Add audio data provider and a simplified DeepSpeech2 model configuration.

### LINKED ISSUE #2226: Add audio data provider and preprocessor for speech recognition datasets.
- Prepare one or more public English speech recognition data sets (e.g. LibriSpeech), and respective baselines.
- Convert all audio file formats to .wav format.
- Add a file manifest generator for each dataset, and add a merger if there exist more than one datasets. Make this interface unified across different datasets.
- Add spectrogram feature extractor, power normalizer etc.
- Add transcription text parser (tokenization, dictionary generation etc).
- Add batch data reader with SortaGrad.
- Refer to the [DS2 design doc](https://github.com/xinghai-sun/Paddle/tree/ds2_design_doc/doc/design/speech) and update it when necessary.
- Please pull your codes and docs into PaddlePaddle/models.

### LINKED ISSUE #2231: Add simplified model configuration for DeepSpeech2.
- Create a simplified model configuration for DeepSpeech2 on V2 APIs.
   - With fixed-length (by padding) batches.
   - With only bidirectional-GRU.
   - With only greedy decoder (max + post-processing).
- Here "simplified" refers to the above limitations, so that we can move on without being blocked by other un-resolved tasks (e.g. [2228](https://github.com/PaddlePaddle/Paddle/issues/2228), [2230](https://github.com/PaddlePaddle/Paddle/issues/2230) etc.).
- About network architecture please refer to [DS2 paper](https://arxiv.org/abs/1512.02595) and the [design doc](https://github.com/xinghai-sun/Paddle/tree/ds2_design_doc/doc/design/speech).
- Please test it on GPUs, and make sure to achieve at least a normal convergence.
- Please pull your codes and docs into PaddlePaddle/models.

### PR DESCRIPTION
**Pull request by xinghai-sun**
resolved [issue 2226](https://github.com/PaddlePaddle/Paddle/issues/2226)  
resolved [issue 2231](https://github.com/PaddlePaddle/Paddle/issues/2231)

### CODE REVIEW

File: deep_speech_2/train.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    # first pass with sortagrad
+    if args.use_sortagrad:
+        trainer.train(
+            reader=train_batch_reader_with_sortagrad,
+            event_handler=event_handler,
+            num_passes=1,
+            feeding=feeding)
+        args.num_passes -= 1
+    # other passes without sortagrad
+    trainer.train(
```
Comment (xinghai-sun) [inline deep_speech_2/train.py:177]: If `args.use_sortagrad` is true, `trainer.train` will be called twice. However, the second `trainer.train` call will get stuck (no progress, no error).  `trainer.train` does not support multiple function calls ?

Comment (xinghai-sun) [reply deep_speech_2/train.py:177]: I think more elaborate controlling interfaces of training process need to be exposed to users for more flexible training flow control. E.g. in this DS2 case, training data needs to be changed during training. In other cases, parts of model needs freezing for a while or being trained alternatively (e.g. GAN).

File: deep_speech_2/model.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            input=last_state,
+            size=size,
+            act=paddle.activation.Linear(),
+            bias_attr=False)
+        return paddle.layer.addto(
+            name=name + "_state", input=[input_fc_bn, state_fc], act=act)
+
+    forward = paddle.layer.recurrent_group(
+        step=__simple_rnn_step__, input=input)
+    return forward
```
Comment (lcy-seso) [inline deep_speech_2/model.py]: This has been fixed. Need to be updated.

File: deep_speech_2/train.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            if event.batch_id % 10 == 0:
+                print "/nPass: %d, Batch: %d, TrainCost: %f" % (
+                    event.pass_id, event.batch_id, event.cost)
+            else:
+                sys.stdout.write('.')
+                sys.stdout.flush()
+        if isinstance(event, paddle.event.EndPass):
+            result = trainer.test(reader=test_batch_reader, feeding=feeding)
+            print "Pass: %d, TestCost: %s" % (event.pass_id, result.cost)
+            with gzip.open("params.tar.gz", 'w') as f:
```
Comment (lcy-seso) [inline deep_speech_2/train.py]: Save the trained model according to the pass number, otherwise, the later saved model overwrites the former ones.

File: deep_speech_2/model.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        act=paddle.activation.Linear(),
+        bias_attr=True)
+    # ctc cost
+    cost = paddle.layer.warp_ctc(
+        input=fc,
+        label=text_data,
+        size=dict_size + 1,
+        blank=dict_size,
+        norm_by_times=True)
+    # max decoder
```
Comment (lcy-seso) [inline deep_speech_2/model.py:134]: If `max_id` is not needed in training, I think it should be put into a  testing branch.

File: deep_speech_2/infer.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    _, max_id = deep_speech2(
+        audio_data=audio_data,
+        text_data=text_data,
+        dict_size=dict_size,
+        num_conv_layers=args.num_conv_layers,
+        num_rnn_layers=args.num_rnn_layers,
+        rnn_size=args.rnn_layer_size)
+
+    # load parameters
+    parameters = paddle.parameters.Parameters.from_tar(
```
Comment (lcy-seso) [inline deep_speech_2/infer.py:103]: Save / Load models according to the pass index.

**Review by lcy-seso — CHANGES_REQUESTED** [review]
almost LGTM.

File: deep_speech_2/librispeech.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import tarfile
+import argparse
+import soundfile
+import json
+
+DATA_HOME = os.path.expanduser('~/.cache/paddle/dataset/speech')
+
+URL_TEST = "http://www.openslr.org/resources/12/test-clean.tar.gz"
+URL_DEV = "http://www.openslr.org/resources/12/dev-clean.tar.gz"
+URL_TRAIN = "http://www.openslr.org/resources/12/train-clean-100.tar.gz"
```
Comment (qingqing01) [inline deep_speech_2/librispeech.py]: Add `MD5` to check data.

File: deep_speech_2/audio_data_utils.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+import numpy as np
+import os
+
+RANDOM_SEED = 0
+logger = logging.getLogger(__name__)
+
+
+class DataGenerator(object):
+    """
+    DataGenerator provides basic audio data preprocessing pipeline, and offer
```
Comment (qingqing01) [inline deep_speech_2/audio_data_utils.py]: offer -> offers

File: deep_speech_2/librispeech.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+def unpack(filepath, target_dir):
+    print("Unpacking %s ..." % filepath)
+    tar = tarfile.open(filepath)
+    tar.extractall(target_dir)
+    tar.close()
+    return target_dir
+
+
+def create_manifest(data_dir, manifest_path):
```
Comment (qingqing01) [inline deep_speech_2/librispeech.py:75]: What manifest mean?  Add some comments.

File: deep_speech_2/model.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        block_x=1,
+        block_y=21)
+    # rnn group
+    rnn_group_output = rnn_group(
+        input=conv2seq, size=rnn_size, num_stacks=num_rnn_layers)
+    # output token distribution
+    fc = paddle.layer.fc(
+        input=rnn_group_output,
+        size=dict_size + 1,
+        act=paddle.activation.Linear(),
```
Comment (qingqing01) [inline deep_speech_2/model.py:125]: Note,  the activation should be `softmax` in inference mode.

File: deep_speech_2/requirements.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+pip install wget
+pip install soundfile
+
+# For Ubuntu only
+apt-get install libsndfile1
```
Comment (qingqing01) [inline deep_speech_2/requirements.sh]: Need to explain in the document.

File: deep_speech_2/infer.py (hunk 5 above)
Comment (xinghai-sun) [reply deep_speech_2/infer.py:103]: Since overfitting rarely happens in DS2, it is not necessary to save multiple models with the pass index. Currently the latest one will be enough. Might add it in the future if necessary.

File: deep_speech_2/audio_data_utils.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    def __init__(self,
+                 vocab_filepath,
+                 normalizer_manifest_path,
+                 normalizer_num_samples=100,
+                 max_duration=20.0,
+                 min_duration=0.0,
+                 stride_ms=10.0,
+                 window_ms=20.0,
+                 max_frequency=None):
+        self.__max_duration__ = max_duration
```
Comment (pkuyym) [inline deep_speech_2/audio_data_utils.py:60]: Python里面最好不要使用__XX__来定义自己的函数和变量，因为__init__、__del__等是python内置的命名方式，下同。如果定义私有的函数或变量只在前面加下划线就可以，印象中__XX是私有的，子类无法访问，_XX也是私有的但是子类可访问

File: deep_speech_2/librispeech.py (hunk 8 above)
Comment (xinghai-sun) [reply deep_speech_2/librispeech.py:75]: Done.

File: deep_speech_2/librispeech.py (hunk 6 above)
Comment (xinghai-sun) [reply deep_speech_2/librispeech.py]: Done.

File: deep_speech_2/model.py (hunk 2 above)
Comment (xinghai-sun) [reply deep_speech_2/model.py]: Done.

File: deep_speech_2/model.py (hunk 9 above)
Comment (xinghai-sun) [reply deep_speech_2/model.py:125]: Current code only contains the Best Path Decoder, which do not require a `softmax` activation.

File: deep_speech_2/audio_data_utils.py [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+        self.__min_duration__ = min_duration
+        self.__stride_ms__ = stride_ms
+        self.__window_ms__ = window_ms
+        self.__max_frequency__ = max_frequency
+        self.__random__ = random.Random(RANDOM_SEED)
```
Comment (pkuyym) [inline deep_speech_2/audio_data_utils.py:65]: 是否考虑开放接口给用户，让其指定随机种子？

File: deep_speech_2/model.py (hunk 4 above)
Comment (xinghai-sun) [reply deep_speech_2/model.py:134]: Will refactor this part later, together with beam search decoder.

File: deep_speech_2/model.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    # ctc cost
+    cost = paddle.layer.warp_ctc(
+        input=fc,
+        label=text_data,
+        size=dict_size + 1,
+        blank=dict_size,
+        norm_by_times=True)
+    # max decoder
+    max_id = paddle.layer.max_id(input=fc)
+    return cost, max_id
```
Comment (pkuyym) [inline deep_speech_2/model.py:136]: 可以根据训练或者预测分别返回cost或max_id，如果是预测的话，cost应该是非必须的吧

File: deep_speech_2/requirements.sh (hunk 10 above)
Comment (pkuyym) [reply deep_speech_2/requirements.sh]: python部分可以改成提供一个 requires.txt

**Review by pkuyym — COMMENTED** [review]
Almost LGTM.

File: deep_speech_2/train.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    "--num_conv_layers", default=3, type=int, help="Convolution layer number.")
+parser.add_argument(
+    "--num_rnn_layers", default=5, type=int, help="RNN layer number.")
+parser.add_argument(
+    "--rnn_layer_size", default=512, type=int, help="RNN layer cell number.")
+parser.add_argument(
+    "--use_gpu", default=True, type=bool, help="Use gpu or not.")
+parser.add_argument(
+    "--use_sortagrad", default=False, type=bool, help="Use sortagrad or not.")
+parser.add_argument(
```
Comment (kuke) [inline deep_speech_2/train.py]: 第20行已经定义了"--trainer"，help信息相同，是否重复定义？

**Review by kuke — COMMENTED** [review]
almost LGTM

File: deep_speech_2/requirements.sh (hunk 10 above)
Comment (xinghai-sun) [reply deep_speech_2/requirements.sh]: Done.

File: deep_speech_2/train.py (hunk 3 above)
Comment (xinghai-sun) [reply deep_speech_2/train.py]: Since overfitting rarely happens in DS2, it is not necessary to save multiple models with the pass index. Currently the latest one will be enough. Might add it in the future if necessary.

**Review by xinghai-sun — COMMENTED** [review]
Done. Thanks for the review!

File: deep_speech_2/train.py (hunk 14 above)
Comment (xinghai-sun) [reply deep_speech_2/train.py]: Yes. Removed.

File: deep_speech_2/audio_data_utils.py (hunk 7 above)
Comment (xinghai-sun) [reply deep_speech_2/audio_data_utils.py]: Done.

File: deep_speech_2/audio_data_utils.py (hunk 11 above)
Comment (xinghai-sun) [reply deep_speech_2/audio_data_utils.py:60]: 目前参考paddle相关python规范如此，建议和大家讨论之后再改？

File: deep_speech_2/audio_data_utils.py (hunk 12 above)
Comment (xinghai-sun) [reply deep_speech_2/audio_data_utils.py:65]: RANDOM_SEED 不重要，不建议开放。

File: deep_speech_2/model.py (hunk 13 above)
Comment (xinghai-sun) [reply deep_speech_2/model.py:136]: Will refactor this part later when merging with beam search decoder.

**Review by xinghai-sun — COMMENTED** [review]
Thanks for the review.

**Comment by xinghai-sun** [comment]
Now the model can run smoothly, with a good convergence and reasonable decoding results.

**Review by lcy-seso — APPROVED** [review]
LGTM
